editor/src/viewport_tools/tool.rs

```diff
@@ -272,3 +272,13 @@ pub fn update_working_colors(document_data: &DocumentToolData, responses: &mut V
 		.into(),
 	);
 }
+
+/// Calculate the path that new layers should be inserted to.
+/// Depends on the selected layers as well as their types (Folder/Non-Folder)
+pub fn get_new_layer_location(document: &DocumentMessageHandler) -> Vec<u64> {
+	// If the selected layers dont actually exist, a new uuid for the
+	// root folder will be returned
+	let mut path = document.graphene_document.shallowest_common_folder(document.selected_layers()).map_or(vec![], |v| v.to_vec());
+	path.push(generate_uuid());
+	path
+}
```

Comment:
I mean you pass it a reference to the DocumentMessageHandler, the most natural place to but it would be there.
I'm not entirely sure what I think about the function name yet, It might be nice if the name somehow contained "path" instead of location, but I'll have to think about that a bit
Other than that this all looks very good!

Reply:
Hmm makes sense. I renamed the function to `get_path_for_new_layer` and moved it into `document_message_handler.rs`.